### PR TITLE
[BUGFIX] Use flag JSON_THROW_ON_ERROR

### PR DESCRIPTION
**Pull request by linawolf**
We do not need to throw an exception ourselves then as json_decode will throw an error now when parsing the json failed

### CODE REVIEW

**Review by brotkrueml — COMMENTED** [review]
Please see: https://github.com/TYPO3-Documentation/t3docs-examples/pull/158
which makes this PR obsolete.

Not throwing an exception using json_decode is intended. In case of an error the result is null which will throw then a custom exception. This is much cleaner and smaller than a try...catch construct.

**Comment by linawolf** [comment]
Throwing the JSON exception is much cleaner: 
The calling code can detect what went wrong and take according actions. 

Swallowing an exception to throw one with less information is usually not good practice and makes debugging very hard. If you do need to  change the error message it should be done right before information is returned to the user, Here it should be done in examples/Classes/Http/MeowInformationRequester.php. the try-catch block is there, anyway.

I would not suggest to change the errormessage here, though as this is a console command and we can expect that developers or admins are running it. They need the complete information.

**Comment by brotkrueml** [comment]
I disagree. When the service is down and returns, for example, a 404 page, for example, then a JSON error is thrown. A technical JSON error should not be given back to a user. The service is not available. Period. If a user is given: "The JSON is not valid" - what does he have to do? He can't fix the response?

Before PHP8 I would have check for null and throw an error or catch the JSON exception and catch it to something meaningful.
A possibility would be to log the original exception message and return a meaningful message to the user. But this is out of scope for this example, it would blow up the example. We want to show how to initiate a HTTP request, not how to catch exceptions and log errors.

**Comment by brotkrueml** [comment]
But we can go the full stack if you want: Inject a logger instance, try...catch the JSON exception, log it and give the user something meaningful. Feel free to change it that way.

**Comment by brotkrueml** [comment]
@linawolf Adjusted the class, have a look.

File: Classes/Http/MeowInformationRequester.php [hunk 1]
```diff
@@ -59,19 +61,10 @@ public function request(): string
...
-            $this->logger->error($e->getMessage());
-            throw new \RuntimeException(
-                'The service did not return a valid response. Try again later.',
-                1666413231,
-                $e
-            );
-        }
+        // Get the content as a string on a successful request
+        $content = $response->getBody()->getContents();
+        $result = json_decode($content, true, flags: JSON_THROW_ON_ERROR)['fact'];
```
Comment (brotkrueml) [inline Classes/Http/MeowInformationRequester.php:65]: That won't work if fact is not returned in response (undefined array key), therefore the `?? throw`syntax here to cache it.

This should not happen (except the service is changing returned format).

**Review by brotkrueml — CHANGES_REQUESTED** [review]

**Comment by linawolf** [comment]
`json_decode ()` returns mixed according to the docs (https://www.php.net/manual/en/function.json-decode.php#refsect1-function.json-decode-returnvalues) 
The method `MeowInformationRequester::request()` returns a string however. So either we change the return value to `mixed` or throw an error if anything but a string is given. Only testing for null is not sufficient.

Regarding the exception handling, I would not return "User friendly" error messages in the Service class but in the command. Here we call the Service from a command so we can expect it is called by an administrator and we can make English error messages. If the same Service method was called from a GUI we might need localized error messages and maybe want to send a flash message or whatever. 

As the command is now the place where we swallow the original error message this is also the place where I would log the error. Bdw. loging has become really easy, now more then two lines of code ;)

File: Classes/Http/MeowInformationRequester.php (hunk 1 above)
Comment (linawolf) [reply Classes/Http/MeowInformationRequester.php:65]: In my daily life I am dealing with services that change their formats and return unexpected stuff without notice all the time. Then the administrator or developer needs to give them a call and say "What the f* have you been doing now". So this information might be important. We cannot rely on a format.

Comment (linawolf) [reply Classes/Http/MeowInformationRequester.php:65]: Oh, ok, the thing about the array key I somehow didn't see

**Comment by brotkrueml** [comment]
Yup, I am fine with the `is_string` check. Mind teh undefined array key above ;-)

File: Classes/Http/MeowInformationRequester.php [hunk 2]
```diff
@@ -57,9 +60,9 @@ public function request(): string
                 'The request did not return JSON data'
             );
         }
-
         // Get the content as a string on a successful request
-        return json_decode($response->getBody()->getContents(), true)['fact']
-            ?? throw new \RuntimeException('No information available');
+        $content = $response->getBody()->getContents();
+        return (string) json_decode($content, true, flags: JSON_THROW_ON_ERROR)['fact']??
+            throw new \RuntimeException('The service returned an unexpected format. ', 1666413230);
```
Comment (brotkrueml) [inline Classes/Http/MeowInformationRequester.php:66]:
```suggestion
            throw new \RuntimeException('The service returned an unexpected format.', 1666413230);
```

File: Classes/Http/MeowInformationRequester.php [hunk 3]
```diff
@@ -30,6 +30,9 @@ public function __construct(
     ) {
     }
 
+    /**
+     * @throws \JsonException
```
Comment (brotkrueml) [inline Classes/Http/MeowInformationRequester.php:34]:
```suggestion
     * @throws \JsonException
     * @throws \RuntimeException
```

**Comment by linawolf** [comment]
wait a moment, running ci just now

**Review by brotkrueml — APPROVED** [review]

**Pull request merged**